test/integration/infrastructure/infrastructure_test.go

```diff
@@ -149,61 +151,143 @@ var _ = Describe("Infrastructure tests", func() {
+
+				By("wait until infrastructure is deleted")
+				// deletion should succeed even though creation failed with invalid credentials (no-op)
+				err := extensions.WaitUntilExtensionObjectDeleted(
+					ctx,
+					c,
+					logger,
+					infra,
+					extensionsv1alpha1.InfrastructureResource,
+					10*time.Second,
+					30*time.Minute,
+				)
+				Expect(err).NotTo(HaveOccurred())
+			}()
+
+			By("create namespace for test execution")
+			namespace = &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "provider-alicloud-test-",
+				},
+			}
+			Expect(c.Create(ctx, namespace)).To(Succeed())
+
+			By("deploy invalid cloudprovider secret into namespace")
+			secret := &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      secretName,
+					Namespace: namespace.Name,
+				},
+				Data: map[string][]byte{
+					alicloud.AccessKeyID:     []byte("invalid"),
+					alicloud.AccessKeySecret: []byte("fake"),
+				},
+			}
+			Expect(c.Create(ctx, secret)).To(Succeed())
+
+			By("create cluster which contains information of shoot info. It is used for encrypted image testing")
+			cluster, err = newCluster(namespace.Name)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(c.Create(ctx, cluster)).To(Succeed())
+
+			By("create infrastructure")
+			infra, err = newInfrastructure(namespace.Name, providerConfig)
 			Expect(err).NotTo(HaveOccurred())
+			Expect(c.Create(ctx, infra)).To(Succeed())
+
+			By("wait until infrastructure creation has failed")
+			err = extensions.WaitUntilExtensionObjectReady(
+				ctx,
+				c,
+				logger,
+				infra,
+				extensionsv1alpha1.InfrastructureResource,
+				10*time.Second,
+				30*time.Second,
+				5*time.Minute,
+				nil,
+			)
+			Expect(err).To(MatchError(ContainSubstring("Specified access key is not found")))
+			var errorWithCode *gardencorev1beta1helper.ErrorWithCodes
+			Expect(errors.As(err, &errorWithCode)).To(BeTrue())
+			Expect(errorWithCode.Codes()).To(ConsistOf(gardencorev1beta1.ErrorInfraUnauthenticated, gardencorev1beta1.ErrorConfigurationProblem))
 		})
 	})
 })
 
 func runTest(ctx context.Context, logger *logrus.Entry, c client.Client, providerConfig *alicloudv1alpha1.InfrastructureConfig, decoder runtime.Decoder, clientFactory alicloudclient.ClientFactory) error {
 	var (
-		infra                     *extensionsv1alpha1.Infrastructure
 		namespace                 *corev1.Namespace
 		cluster                   *extensionsv1alpha1.Cluster
+		infra                     *extensionsv1alpha1.Infrastructure
 		infrastructureIdentifiers infrastructureIdentifiers
 		err                       error
 	)
 
-	var cleanupHandle framework.CleanupActionHandle
-	cleanupHandle = framework.AddCleanupAction(func() {
+	framework.AddCleanupAction(func() {
+		By("cleaning up namespace and cluster")
+		Expect(client.IgnoreNotFound(c.Delete(ctx, namespace))).To(Succeed())
+		Expect(client.IgnoreNotFound(c.Delete(ctx, cluster))).To(Succeed())
+	})
+
+	defer func() {
```

Comment:
If secret is rotated and not valid anymore, infra won't be deleted.  Then it's still not possible to remove namespace even you move the cleanup to aftereachsuite. Am I right?

Reply:
Yes. If secret becomes invalid during the test, then the test here will not succeed in any case. Should that happen, currently we can only manually update the credentials in the TM cluster.